src/node/nodestate.h

```diff
@@ -408,12 +420,31 @@ namespace ccf
 #endif
       join_rpc.params.quote = quote;
 
-      auto join_req =
-        jsonrpc::pack(nlohmann::json(join_rpc), jsonrpc::Pack::MsgPack);
+      LOG_DEBUG_FMT(
+        "Sending join request to {}:{}",
+        args.config.joining.target_host,
+        args.config.joining.target_port);
+
+      join_client->send(jsonrpc::pack(join_rpc, jsonrpc::Pack::MsgPack));
+    }
+
+    void join(const Join::In& args)
+    {
+      std::lock_guard<SpinLock> guard(lock);
+      sm.expect(State::pending);
 
-      LOG_INFO_FMT("Sending join request");
+      initiate_join(args);
 
-      join_client->send(join_req);
+      using namespace std::chrono_literals;
+      join_timer = timers.new_timer(1s, [this, args]() {
+        if (sm.check(State::pending))
```

Comment:
1. Do we want to react if state is set to RETIRED if a node is rejected?
2. Does it know to stop if the node(s) that it communicates with die?

Reply:
I believe this is addressed as part of https://github.com/microsoft/CCF/issues/397 and https://github.com/microsoft/CCF/issues/398, respectively.